src/ray/rpc/client_call.h

```diff
@@ -303,27 +272,27 @@ class ClientCallManager {
       } else if (status != grpc::CompletionQueue::TIMEOUT) {
         // NOTE: CompletionQueue::TIMEOUT and gRPC deadline exceeded are different.
         // If the client deadline is exceeded, event is obtained at this block.
-        auto tag = reinterpret_cast<ClientCallTag *>(got_tag);
-        // Refresh the tag.
-        got_tag = nullptr;
-        tag->GetCall()->SetReturnStatus();
-        std::shared_ptr<StatsHandle> stats_handle = tag->GetCall()->GetStatsHandle();
+
+        auto call = static_cast<ClientCall *>(got_tag);
+        call->SetReturnStatus();
+        std::shared_ptr<StatsHandle> stats_handle = call->GetStatsHandle();
         RAY_CHECK_NE(stats_handle, nullptr);
+
         if (ok && !main_service_.stopped() && !shutdown_) {
           // Post the callback to the main event loop.
           main_service_.post(
-              [tag]() {
-                tag->GetCall()->OnReplyReceived();
+              [call]() {
+                call->OnReplyReceived();
                 // The call is finished, and we can delete this tag now.
-                delete tag;
+                delete call;
```

Comment:
I'd really prefer avoid handling naked manually allocated pointers across the single method boundaries and instead

1. Keep `ClientCallTag`
2. Copy shared_ptr to ClientCall *and* delete tag as soon as we dequeue from CQ